File: Linked-List/dll.cpp
#include <iostream>
using namespace std;

template <typename T> struct Node {
  T val;
  Node *prev;
  Node *next;
};

template <typename T> class DLL {
private:
  Node<T> *header;
  Node<T> *trailer;

public:
  DLL() {
    header = new Node<T>;
    trailer = new Node<T>;
    header->next = trailer;
    trailer->prev = header;
  }

  ~DLL() {
    while (!empty()) {
      removeFront();
    }
    delete header;
    delete trailer;
  }

  bool empty() const { return (header->next == trailer); }

  const T &front() const { return (header->next->val); }

  const T &back() const { return (trailer->prev->val); }

  void addFront(const T &data) { add(data, header->next); }

  void addBack(const T &data) { add(data, trailer); }

  // adds a new node before the target node
  void add(const T &data, Node<T> *target) {
    Node<T> *newNode = new Node<T>;
    newNode->val = data;
    newNode->next = target;
    newNode->prev = target->prev;
    target->prev->next = newNode;
    target->prev = newNode;
  }

  void remove(Node<T> *target) {
    Node<T> *pred = target->prev;
    Node<T> *succ = target->next;
    pred->next = succ;
    succ->prev = pred;
    delete target;
  }

  void removeFront() { remove(header->next); }

  void removeBack() { remove(trailer->prev); }

  void reverse() {
    Node<T> *temp = nullptr;
    Node<T> *current = header;
    while (current != nullptr) {
      temp = current->prev;
      current->prev = current->next;
      current->next = temp;
      current = current->prev;
    }
    temp = header;
    header = trailer;
    trailer = temp;
    header->prev = nullptr;
    trailer->next = nullptr;
  }
};

int main() {}

File: Linked-List/dll.java
public class dll<E>{
    private static class Node<E>{
        private E element; 
        private Node<E> prev; 
        private Node<E> next; 
        public Node(E e, Node<E> p, Node<E> n){
            element = e; 
            prev = p;
            next = n; 
        }
        
        public E getElement() {return element;}
        public Node<E> getNext() {return next;}
        public Node<E> getPrev() {return prev;}
        public void setPrev(Node<E> p) {prev = p;}
        public void setNext(Node<E> n) {next = n;}

    }

    private Node<E> header; 
    private Node<E> trailer; 
    private int size = 0; 
    public dll(){
        header = new Node<>(null,null,null); 
        trailer = new Node<>(null, header, null); 
        header.setNext(trailer); 
    }
    
    public int size() {return size;}
    public boolean isEmpty() {return size == 0;}

    public E first(){
        if (isEmpty()) {return null;}
        return header.getNext().getElement(); 
    }
    
    public E last() {
        if (isEmpty()) return null; 
        return trailer.getPrev().getElement();
    }

    public void addFirst(E e){
        addBetween(e, header, header.getNext()); 
    }

    public void addLast(E e){
        addBetween(e, trailer.getPrev(), trailer); 
    }

    private void addBetween(E e, Node<E> first, Node<E> second){
        Node<E> newOne = new Node<> (e, first, second); 
        first.setNext((newOne));
        second.setPrev(newOne);
    }
    
    private E remove(Node<E> node){ 
        Node<E> prev = node.getPrev(); 
        Node<E> next = node.getNext(); 
        prev.setNext(next); 
        next.setPrev(prev); 
        size--; 
        return node.getElement();
    }

    public void reverse(){
        Node<E> current = header;
        Node<E> next; 
        Node<E> prev = null;  
        while (current != null){
            current.prev = prev; 
            prev.next = current; 
            prev = current; 
            current = current.getNext(); 


        } 
        prev = header; 
    }
}